File: 417-pacific-atlantic-water-flow/pacific-atlantic-water-flow.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        row = len(heights)
        col = len(heights[0])
        pac = set()
        atl = set()
        result = []

        def dfs(r, c, visit, pastHeight):
            if r < 0 or c < 0 or r == row or c == col or pastHeight>heights[r][c] or (r, c) in visit: return
            visit.add((r, c))
            dfs(r + 1, c, visit, heights[r][c])
            dfs(r - 1, c, visit, heights[r][c])
            dfs(r, c + 1, visit, heights[r][c])
            dfs(r, c - 1, visit, heights[r][c])

        
        for i in range(col):
            dfs(0, i, pac, heights[0][i])
            dfs(row-1, i, atl, heights[row-1][i])
        
        for i in range(row):
            dfs(i, 0, pac, heights[i][0])
            dfs(i, col - 1, atl, heights[i][col - 1])
        
        for i in range(row):
            for j in range(col):
                if (i, j) in pac and (i, j) in atl:
                    result.append([i, j])

        return result

File: 417-pacific-atlantic-water-flow/pacific-atlantic-water-flow.java
class Solution {
    int row;
    int col;

    void dfs(int r, int c, List<Pair<Integer, Integer>> visited, int prevHeight, int[][] heights){
        if (r < 0 || c < 0 || r == row || c == col || prevHeight > heights[r][c] || visited.contains(new Pair<Integer, Integer>(r, c))) return;
        visited.add(new Pair<Integer, Integer>(r, c));
        dfs(r + 1, c, visited, heights[r][c], heights);
        dfs(r - 1, c, visited, heights[r][c], heights);
        dfs(r, c + 1, visited, heights[r][c], heights);
        dfs(r, c - 1, visited, heights[r][c], heights);
    }

    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        row = heights.length;
        col = heights[0].length;
        List<Pair<Integer, Integer>> pac = new ArrayList<Pair<Integer, Integer>>();
        List<Pair<Integer, Integer>> alt = new ArrayList<Pair<Integer, Integer>>();
        List<List<Integer>> result = new ArrayList<List<Integer>>();

        for (int i = 0; i < col; i++){
            dfs(0, i, pac, heights[0][i], heights);
            dfs(row - 1, i, alt, heights[row-1][i], heights);
        }

        for (int i = 0; i < row; i++){
            dfs(i, 0, pac, heights[i][0], heights);
            dfs(i, col - 1, alt, heights[i][col - 1], heights);
        }

        for (int i = 0; i < row; i++){
            for (int j = 0; j < col; j++){
                if (pac.contains(new Pair<Integer, Integer>(i,j)) && alt.contains(new Pair<Integer, Integer>(i,j))){
                    result.add(Arrays.asList(i, j));
                }
            }
        }

        return result;
    }
}